runtime/src/consensus/registry.rs

```diff
@@ -18,6 +18,211 @@ use crate::{
+    pub version: Version,
+
+    /// Node's capabilities for a given runtime.
+    pub capabilities: Capabilities,
+
+    /// Extra per node + per runtime opaque data associated with the current instance.
+    pub extra_info: Option<Vec<u8>>,
+}
+
+/// TEE hardware implementation.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, cbor::Encode, cbor::Decode)]
+#[cbor(transparent)]
+pub struct RolesMask(pub u32);
+
+// XXX: Would be nicer to use bitflags crate for this, but there is no way to add
+// custom derives to the enum at the moment.
+// Should be possible in v2.0 (https://github.com/bitflags/bitflags/issues/262).
+impl RolesMask {
+    /// Empty roles mask.
+    pub const ROLE_EMPTY: RolesMask = RolesMask(0);
+    /// Compute worker role.
+    pub const ROLE_COMPUTE_WORKER: RolesMask = RolesMask(1 << 0);
+    /// Reserved role (storage role in v1 descriptors).
+    pub const ROLE_RESERVED_2: RolesMask = RolesMask(1 << 1);
+    /// Key manager role.
+    pub const ROLE_KEY_MANAGER: RolesMask = RolesMask(1 << 2);
+    /// Validator role.
+    pub const ROLE_VALIDATOR: RolesMask = RolesMask(1 << 3);
+    /// Public consensus RPC services worker role.
+    pub const ROLE_CONSENSUS_RPC: RolesMask = RolesMask(1 << 4);
+    /// Public storage RPC services worker role.
+    pub const ROLE_STORAGE_RPC: RolesMask = RolesMask(1 << 5);
+
+    // Bits of the Oasis node roles bitmask that are reserved and must not be used.
+    pub const ROLES_RESERVED: RolesMask =
+        RolesMask(u32::MAX & !((Self::ROLE_STORAGE_RPC.0 << 1) - 1) | Self::ROLE_RESERVED_2.0);
+}
+
+impl PartialOrd for RolesMask {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl Ord for RolesMask {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl Default for RolesMask {
+    fn default() -> Self {
+        Self::ROLE_EMPTY
+    }
+}
+
+/// Node registry descriptor.
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, cbor::Encode, cbor::Decode)]
+pub struct Node {
+    /// Structure version.
+    pub v: u16,
+
+    /// Public key identifying the node.
+    pub id: PublicKey,
+
+    /// Public key identifying the Entity controlling the node.
+    pub entity_id: PublicKey,
+
+    /// Epoch in which the node's commitment expires.
+    pub expiration: u64,
+
+    /// Information for connecting to this node via TLS.
+    pub tls: TLSInfo,
+
+    /// Information for connecting to this node via P2P.
+    pub p2p: P2PInfo,
+
+    /// Information for connecting to this node as a consensus member.
+    pub consensus: ConsensusInfo,
+
+    /// Information for this node's participation in VRF based elections.
+    #[cbor(optional, default)]
+    pub vrf: Option<VRFInfo>,
+
+    /// Information for this node's participation in the old PVSS based random beacon protocol.
+    /// Deprecated.
+    #[cbor(optional, default)]
+    pub beacon: Option<Vec<u8>>,
```

Comment:
Sigh, there's a PR that gets rid of this on the Go side, but apparently "no breaking changes" or something, so the branch is rotting.

https://github.com/oasisprotocol/oasis-core/pull/4668